examples/scene/bsn.rs

```diff
@@ -0,0 +1,197 @@
+        Name("Base")
+        Sprite {
+            handle: "asset://branding/bevy_bird_dark.png",
+            size: 1,
+            nested: Nested {
+                handle: @"asset://hello.png"
+            }
+        }
+        Transform::from_translation(Vec3::new(1.0, 1.0, 1.0))
+        Team::Red {
+            x: 10,
+            y: Nested {
+                foo: 10
+            },
+        }
+        Gen::<usize> {
+            value: 10,
+        }
+        on(|event: On<Explode>| {
+        })
+        Foo(100, @"asset://branding/bevy_bird_dark.png")
+        [
+            (:sprite_big Sprite { size: 2 }),
+            :widget(bsn_list![Text::new("hi")]),
+            {sprites},
+        ]
+    }
+}
+
+fn sprite_big() -> impl Scene {
+    bsn! {
+        Sprite { size: 100000, handle: "asset://branding/icon.png" }
+    }
+}
+
+fn x() -> impl Scene {
+    bsn! {
+        :"scene://transform_1000.bsn"
+        Transform { translation: Vec3 { x: 11.0 } }
+    }
+}
+
+fn transform_1000() -> impl Scene {
+    bsn! {
+        Transform {
+            translation: Vec3 { x: 1000.0, y: 1000.0 }
+        }
+    }
+}
+
+fn transform_1337() -> impl Scene {
+    bsn! {
+        Transform {
+            translation: Vec3 { x: 1337.0 }
+        }
+    }
+}
+
+#[derive(Component, Debug, GetTemplate)]
+struct Sprite {
+    handle: Handle<Image>,
+    size: usize,
+    entity: Entity,
+    nested: Nested,
+}
+
+#[derive(Component, Debug, GetTemplate)]
+struct Gen<T: GetTemplate<Template: Default + Template<Output = T>>> {
+    size: usize,
+    value: T,
+}
+
+#[derive(Clone, Debug, GetTemplate)]
+struct Nested {
+    foo: usize,
+    #[template]
+    handle: Handle<Image>,
+}
+
+#[derive(Component, Clone, Debug, GetTemplate)]
+struct Foo(usize, #[template] Handle<Image>);
+
+#[derive(Event, EntityEvent)]
+struct Explode;
+
+#[derive(Component, Default, Clone)]
+struct Thing;
+
+fn print<C: Component + std::fmt::Debug>(
+    query: Query<(Entity, Option<&Name>, Option<&ChildOf>, &C), Changed<C>>,
+) {
+    for (e, name, child_of, c) in &query {
+        println!("Changed {e:?} {name:?} {child_of:?} {c:#?}");
+    }
+}
+
+#[derive(Component, Debug, GetTemplate)]
+enum Team {
+    Red {
+        x: usize,
+        y: Nested,
+    },
+    Blue,
+    #[default]
+    Green(usize, usize),
+}
+
+#[derive(Component, GetTemplate)]
+enum Blah {
+    #[default]
+    A(Hi),
+    B(Arrrrg),
+    C(String),
+}
+
+#[derive(Default)]
+struct Arrrrg(Option<Box<dyn Fn(&mut World) + Send + Sync>>);
+
+impl Clone for Arrrrg {
+    fn clone(&self) -> Self {
+        Self(None)
+    }
+}
+
+impl<F: Fn(&mut World) + Send + Sync> From<F> for Arrrrg {
+    fn from(value: F) -> Self {
+        todo!()
+    }
+}
+
+#[derive(Clone, Default)]
+struct Hi {
+    size: usize,
+}
+
+fn test() -> impl Scene {
+    bsn! {
+        Blah::A(Hi {size: 10})
+        Blah::B(|world: &mut World| {})
+        Blah::C("hi")
+    }
+}
+
+fn widget(children: impl SceneList) -> impl Scene {
```

Comment:
impl Scene, or T: Scene works well for functions that return new scenes.

What are the options for storing a type erased scene/template(?) as a value?

E.g. I could see a use case where a plugin asks the user to provide something like a `ResourceThing(Box<dyn Scene>)`, and then the plugin can use it to spawn scenes when needed.

Reply:
`Scene` is object safe, so you can just box it directly. We could consider implementing Scene for `Box<dyn Scene>` if we decide that is necessary. In userspace, you can also already build a `impl Scene for BoxedScene` wrapper.

Templates are not object safe. There is an `ErasedTemplate` type used by `ResolvedScene`, but it is not usable as a Template directly. In theory it could be framed as a `Template<Output = ()>` though, if we decide that would be useful.

Reply:
This is a *long* standing request of mine, see https://github.com/bevyengine/bevy/issues/3227. I would be extremely happy with a resolution in some form here.